File: scripts/compare_builds.py
"""
PoE2 .build file comparator.
Compares two poe.ninja .build JSON files and shows advantages/disadvantages.

Usage:
    python compare_builds.py build1.build build2.build
"""
import json
import sys
import re
from pathlib import Path
from collections import defaultdict

ASCENDANCY_NAMES = {
    "Witch2": "Blood Mage",
    "Witch1": "Infernalist",
    "Witch3": "Stormweaver",
    "Sorcerer1": "Stormweaver",
    "Sorcerer2": "Chronomancer",
    "Monk1": "Invoker",
    "Monk2": "Acolyte of Chayula",
    "Mercenary1": "Witchhunter",
    "Mercenary2": "Gemlin Legionnaire",
    "Mercenary3": "Tactician",
    "Warrior1": "Warbringer",
    "Warrior2": "Titan",
    "Warrior3": "Smith of Kitava",
    "Ranger1": "Deadeye",
    "Ranger2": "Pathfinder",
    "Huntress1": "Ritualist",
    "Huntress2": "Amazon",
}


def categorize_passive(node_id):
    if node_id.startswith("Ascendancy"):
        return "Ascendancy"
    if node_id.startswith("passive_keystone"):
        return "Keystone"
    if node_id.startswith("jewel_slot"):
        return "Jewel Socket"
    if node_id.startswith("witch_sorceress"):
        return "Witch/Sorceress"
    if node_id.startswith("criticals") or node_id.startswith("critical"):
        return "Critical"
    if node_id.startswith("spell_criticals") or node_id.startswith("spell_critical"):
        return "Spell Critical"
    if node_id.startswith("cast_speed"):
        return "Cast Speed"
    if node_id.startswith("spells") or node_id.startswith("spell"):
        return "Spell"
    if node_id.startswith("projectile_spells") or node_id.startswith("projectile_spell"):
        return "Projectile Spell"
    if node_id.startswith("area_spells") or node_id.startswith("area_attacks") or node_id.startswith("area"):
        return "Area"
    if node_id.startswith("elemental"):
        return "Elemental"
    if node_id.startswith("energy_shield"):
        return "Energy Shield"
    if node_id.startswith("intelligence"):
        return "Intelligence"
    if node_id.startswith("dexterity"):
        return "Dexterity"
    if node_id.startswith("strength"):
        return "Strength"
    if node_id.startswith("attributes"):
        return "Attributes"
    if node_id.startswith("duration"):
        return "Duration"
    return "Other"


def gem_short_name(gem_id):
    match = re.search(r'SkillGem([\w_]+)$', gem_id) or \
            re.search(r'SupportGem([\w_]+)$', gem_id) or \
            re.search(r'/([\w_]+)$', gem_id)
    name = match.group(1) if match else gem_id
    name = re.sub(r'(.)([A-Z])', r'\1 \2', name)
    name = re.sub(r'_', ' ', name)
    name = name.replace(' Two ', ' 2').strip()
    return name


def parse_build(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        data = json.load(f)

    build = {
        'name': data.get('name', 'Unnamed'),
        'author': data.get('author', 'Unknown'),
        'ascendancy': ASCENDANCY_NAMES.get(data.get('ascendancy', ''), data.get('ascendancy', 'Unknown')),
        'ascendancy_id': data.get('ascendancy', ''),
        'passives': [],
        'passives_by_category': defaultdict(list),
        'skills': [],
    }

    for p in data.get('passives', []):
        cat = categorize_passive(p['id'])
        entry = {'id': p['id'], 'weapon_set': p.get('weapon_set', 0), 'category': cat}
        build['passives'].append(entry)
        build['passives_by_category'][cat].append(entry)

    for s in data.get('skills', []):
        skill = {
            'id': s['id'],
            'name': gem_short_name(s['id']),
            'supports': [{'id': ss['id'], 'name': gem_short_name(ss['id'])} for ss in s.get('support_skills', [])]
        }
        build['skills'].append(skill)

    return build


def compare_builds(build_a, build_b):
    diff = {
        'buildA': {'name': build_a['name'], 'ascendancy': build_a['ascendancy'], 'author': build_a['author']},
        'buildB': {'name': build_b['name'], 'ascendancy': build_b['ascendancy'], 'author': build_b['author']},
        'passives': {'onlyA': [], 'onlyB': [], 'shared': []},
        'passives_by_category': defaultdict(lambda: {'onlyA': [], 'onlyB': [], 'shared': []}),
        'skills': {'onlyA': [], 'onlyB': [], 'shared': []},
    }

    passives_a = {p['id']: p for p in build_a['passives']}
    passives_b = {p['id']: p for p in build_b['passives']}

    all_categories = set()
    for p in build_a['passives']:
        all_categories.add(p['category'])
    for p in build_b['passives']:
        all_categories.add(p['category'])

    for cat in all_categories:
        for p in build_a['passives']:
            if p['category'] != cat:
                continue
            target = diff['passives']['shared'] if p['id'] in passives_b else diff['passives']['onlyA']
            target.append({'id': p['id'], 'weapon_set': p['weapon_set']})
            target_cat = diff['passives_by_category'][cat]['shared'] if p['id'] in passives_b else diff['passives_by_category'][cat]['onlyA']
            target_cat.append({'id': p['id'], 'weapon_set': p['weapon_set']})

        for p in build_b['passives']:
            if p['category'] != cat or p['id'] in passives_a:
                continue
            diff['passives']['onlyB'].append({'id': p['id'], 'weapon_set': p['weapon_set']})
            diff['passives_by_category'][cat]['onlyB'].append({'id': p['id'], 'weapon_set': p['weapon_set']})

    # Skills comparison
    skills_a = {}
    for s in build_a['skills']:
        key = re.search(r'/([\w_]+)$', s['id'])
        skills_a[key.group(1) if key else s['id']] = s

    skills_b = {}
    for s in build_b['skills']:
        key = re.search(r'/([\w_]+)$', s['id'])
        skills_b[key.group(1) if key else s['id']] = s

    for s in build_a['skills']:
        key = re.search(r'/([\w_]+)$', s['id'])
        key = key.group(1) if key else s['id']
        if key in skills_b:
            shared = {'name': s['name'], 'id': s['id'], 'supports': {'onlyA': [], 'onlyB': [], 'shared': []}}
            sup_a = {re.search(r'/([\w_]+)$', ss['id']).group(1) if re.search(r'/([\w_]+)$', ss['id']) else ss['id']: ss for ss in s['supports']}
            sup_b = {re.search(r'/([\w_]+)$', ss['id']).group(1) if re.search(r'/([\w_]+)$', ss['id']) else ss['id']: ss for ss in skills_b[key]['supports']}
            for ss in s['supports']:
                skey = re.search(r'/([\w_]+)$', ss['id'])
                skey = skey.group(1) if skey else ss['id']
                if skey in sup_b:
                    shared['supports']['shared'].append(ss['name'])
                else:
                    shared['supports']['onlyA'].append(ss['name'])
            for ss in skills_b[key]['supports']:
                skey = re.search(r'/([\w_]+)$', ss['id'])
                skey = skey.group(1) if skey else ss['id']
                if skey not in sup_a:
                    shared['supports']['onlyB'].append(ss['name'])
            diff['skills']['shared'].append(shared)
        else:
            diff['skills']['onlyA'].append(s)

    for s in build_b['skills']:
        key = re.search(r'/([\w_]+)$', s['id'])
        key = key.group(1) if key else s['id']
        if key not in skills_a:
            diff['skills']['onlyB'].append(s)

    diff['summary'] = {
        'total_passives_a': len(build_a['passives']),
        'total_passives_b': len(build_b['passives']),
        'shared_passives': len(diff['passives']['shared']),
        'unique_passives_a': len(diff['passives']['onlyA']),
        'unique_passives_b': len(diff['passives']['onlyB']),
        'total_skills_a': len(build_a['skills']),
        'total_skills_b': len(build_b['skills']),
        'shared_skills': len(diff['skills']['shared']),
        'unique_skills_a': len(diff['skills']['onlyA']),
        'unique_skills_b': len(diff['skills']['onlyB']),
    }

    return diff


def print_diff(diff):
    s = diff['summary']
    print()
    print("=" * 60)
    print("  BUILD COMPARISON")
    print("=" * 60)
    print()
    print(f"  Build A: {diff['buildA']['name']} ({diff['buildA']['ascendancy']})")
    print(f"  Build B: {diff['buildB']['name']} ({diff['buildB']['ascendancy']})")
    print()
    print(f"  {'─' * 56}")
    print(f"  {'SUMMARY':^56}")
    print(f"  {'─' * 56}")
    print(f"  {'':20} {'A':>8} {'B':>8} {'Shared':>8} {'Unique A':>8} {'Unique B':>8}")
    print(f"  {'Passives':20} {s['total_passives_a']:>8} {s['total_passives_b']:>8} {s['shared_passives']:>8} {s['unique_passives_a']:>8} {s['unique_passives_b']:>8}")
    print(f"  {'Skills':20} {s['total_skills_a']:>8} {s['total_skills_b']:>8} {s['shared_skills']:>8} {s['unique_skills_a']:>8} {s['unique_skills_b']:>8}")
    print()

    if s['shared_passives'] > 0:
        print(f"  SHARED PASSIVES ({s['shared_passives']})")
        for p in diff['passives']['shared']:
            ws = f" (WS{p['weapon_set']})" if p['weapon_set'] > 0 else ""
            print(f"    {p['id']}{ws}")
        print()

    def print_passive_section(items, label):
        if not items:
            return
        by_cat = defaultdict(list)
        for p in items:
            by_cat[categorize_passive(p['id'])].append(p)
        print(f"  {label}")
        for cat in sorted(by_cat.keys()):
            print(f"    [{cat}]")
            for p in by_cat[cat]:
                ws = f" (WS{p['weapon_set']})" if p['weapon_set'] > 0 else ""
                print(f"      {p['id']}{ws}")
        print()

    print_passive_section(diff['passives']['onlyA'], f"UNIQUE TO BUILD A ({diff['buildA']['name']})")
    print_passive_section(diff['passives']['onlyB'], f"UNIQUE TO BUILD B ({diff['buildB']['name']})")

    if s['shared_skills'] > 0:
        print(f"  SHARED SKILLS ({s['shared_skills']})")
        for sk in diff['skills']['shared']:
            print(f"    {sk['name']}")
            sup = sk['supports']
            if sup['shared']:
                print(f"      Supports: {', '.join(sup['shared'])}")
            if sup['onlyA']:
                print(f"      Only in A: {', '.join(sup['onlyA'])}")
            if sup['onlyB']:
                print(f"      Only in B: {', '.join(sup['onlyB'])}")
        print()

    def print_skill_section(items, label):
        if not items:
            return
        print(f"  {label}")
        for sk in items:
            print(f"    {sk['name']}")
            if sk.get('supports'):
                names = [ss['name'] for ss in sk['supports']]
                print(f"      Supports: {', '.join(names)}")
        print()

    print_skill_section(diff['skills']['onlyA'], f"SKILLS ONLY IN BUILD A ({diff['buildA']['name']})")
    print_skill_section(diff['skills']['onlyB'], f"SKILLS ONLY IN BUILD B ({diff['buildB']['name']})")


def main():
    if len(sys.argv) != 3:
        print("Usage: python compare_builds.py <build1.build> <build2.build>")
        sys.exit(1)

    try:
        build_a = parse_build(sys.argv[1])
        build_b = parse_build(sys.argv[2])
        diff = compare_builds(build_a, build_b)
        print_diff(diff)
    except FileNotFoundError as e:
        print(f"Error: File not found - {e}")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON - {e}")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/compare_builds.mjs
#!/usr/bin/env node
/**
 * PoE2 .build file comparator
 * Compares two poe.ninja .build JSON files and shows advantages/disadvantages.
 *
 * Usage: node compare_builds.mjs <build1.build> <build2.build>
 */
import { readFileSync } from 'fs';

const ASCENDANCY_NAMES = {
  Witch2: 'Blood Mage', Witch1: 'Infernalist', Witch3: 'Stormweaver',
  Sorcerer1: 'Stormweaver', Sorcerer2: 'Chronomancer', Sorcerer3: 'N/A',
  Monk1: 'Invoker', Monk2: 'Acolyte of Chayula', Monk3: 'N/A',
  Mercenary1: 'Witchhunter', Mercenary2: 'Gemlin Legionnaire', Mercenary3: 'Tactician',
  Warrior1: 'Warbringer', Warrior2: 'Titan', Warrior3: 'Smith of Kitava',
  Ranger1: 'Deadeye', Ranger2: 'Pathfinder', Ranger3: 'N/A',
  Huntress1: 'Ritualist', Huntress2: 'Amazon', Huntress3: 'N/A',
};

function categorizePassive(id) {
  if (id.startsWith('Ascendancy')) return 'Ascendancy';
  if (id.startsWith('passive_keystone')) return 'Keystone';
  if (id.startsWith('jewel_slot')) return 'Jewel Socket';
  if (id.startsWith('witch_sorceress')) return 'Witch/Sorceress';
  if (id.startsWith('critical')) return 'Critical';
  if (id.startsWith('spell_critical')) return 'Spell Critical';
  if (id.startsWith('cast_speed')) return 'Cast Speed';
  if (id.startsWith('projectile_spell')) return 'Projectile Spell';
  if (id.startsWith('area_spell') || id.startsWith('area_attack') || id.startsWith('area_')) return 'Area';
  if (id.startsWith('spell')) return 'Spell';
  if (id.startsWith('elemental')) return 'Elemental';
  if (id.startsWith('energy_shield')) return 'Energy Shield';
  if (id.startsWith('intelligence')) return 'Intelligence';
  if (id.startsWith('dexterity')) return 'Dexterity';
  if (id.startsWith('strength')) return 'Strength';
  if (id.startsWith('attribute')) return 'Attributes';
  if (id.startsWith('duration')) return 'Duration';
  return 'Other';
}

function gemShortName(gemId) {
  const m = gemId.match(/SkillGem([\w_]+)$/) || gemId.match(/SupportGem([\w_]+)$/) || gemId.match(/\/([\w_]+)$/);
  let name = m ? m[1] : gemId;
  name = name.replace(/([a-z])([A-Z])/g, '$1 $2').replace(/_/g, ' ').replace(/ Two /g, ' 2 ').trim();
  return name;
}

function parseBuild(filepath) {
  const data = JSON.parse(readFileSync(filepath, 'utf-8'));
  const build = { name: data.name || 'Unnamed', author: data.author || 'Unknown',
    ascendancy: ASCENDANCY_NAMES[data.ascendancy] || data.ascendancy || 'Unknown',
    passives: [], passivesByCategory: {}, skills: [] };

  for (const p of data.passives || []) {
    const cat = categorizePassive(p.id);
    const entry = { id: p.id, weaponSet: p.weapon_set || 0, category: cat };
    build.passives.push(entry);
    (build.passivesByCategory[cat] ??= []).push(entry);
  }

  for (const s of data.skills || []) {
    build.skills.push({
      id: s.id, name: gemShortName(s.id),
      supports: (s.support_skills || []).map(ss => ({ id: ss.id, name: gemShortName(ss.id) }))
    });
  }
  return build;
}

function compareBuilds(a, b) {
  const diff = {
    buildA: { name: a.name, ascendancy: a.ascendancy, author: a.author },
    buildB: { name: b.name, ascendancy: b.ascendancy, author: b.author },
    passives: { onlyA: [], onlyB: [], shared: [] },
    passivesByCategory: {},
    skills: { onlyA: [], onlyB: [], shared: [] },
  };

  const setA = new Set(a.passives.map(p => p.id));
  const setB = new Set(b.passives.map(p => p.id));

  for (const p of a.passives) {
    const target = setB.has(p.id) ? diff.passives.shared : diff.passives.onlyA;
    target.push({ id: p.id, weaponSet: p.weaponSet });
  }
  for (const p of b.passives) {
    if (!setA.has(p.id)) diff.passives.onlyB.push({ id: p.id, weaponSet: p.weaponSet });
  }

  // by-category
  const cats = new Set([...Object.keys(a.passivesByCategory), ...Object.keys(b.passivesByCategory)]);
  for (const cat of cats) {
    const ca = new Set((a.passivesByCategory[cat] || []).map(p => p.id));
    const cb = new Set((b.passivesByCategory[cat] || []).map(p => p.id));
    diff.passivesByCategory[cat] = { onlyA: [], onlyB: [], shared: [] };
    for (const p of a.passivesByCategory[cat] || []) {
      (cb.has(p.id) ? diff.passivesByCategory[cat].shared : diff.passivesByCategory[cat].onlyA)
        .push({ id: p.id, weaponSet: p.weaponSet });
    }
    for (const p of b.passivesByCategory[cat] || []) {
      if (!ca.has(p.id)) diff.passivesByCategory[cat].onlyB.push({ id: p.id, weaponSet: p.weaponSet });
    }
  }

  // Skills
  const keyA = {}, keyB = {};
  for (const s of a.skills) { const k = s.id.match(/\/([\w_]+)$/)?.[1] || s.id; keyA[k] = s; }
  for (const s of b.skills) { const k = s.id.match(/\/([\w_]+)$/)?.[1] || s.id; keyB[k] = s; }

  for (const s of a.skills) {
    const k = s.id.match(/\/([\w_]+)$/)?.[1] || s.id;
    if (keyB[k]) {
      const shared = { name: s.name, id: s.id, supports: { onlyA: [], onlyB: [], shared: [] } };
      const supA = {}, supB = {};
      for (const ss of s.supports) { const sk = ss.id.match(/\/([\w_]+)$/)?.[1] || ss.id; supA[sk] = ss; }
      for (const ss of keyB[k].supports) { const sk = ss.id.match(/\/([\w_]+)$/)?.[1] || ss.id; supB[sk] = ss; }
      for (const ss of s.supports) {
        const sk = ss.id.match(/\/([\w_]+)$/)?.[1] || ss.id;
        (supB[sk] ? shared.supports.shared : shared.supports.onlyA).push(ss.name);
      }
      for (const ss of keyB[k].supports) {
        const sk = ss.id.match(/\/([\w_]+)$/)?.[1] || ss.id;
        if (!supA[sk]) shared.supports.onlyB.push(ss.name);
      }
      diff.skills.shared.push(shared);
    } else diff.skills.onlyA.push(s);
  }
  for (const s of b.skills) {
    const k = s.id.match(/\/([\w_]+)$/)?.[1] || s.id;
    if (!keyA[k]) diff.skills.onlyB.push(s);
  }

  diff.summary = {
    totalPassivesA: a.passives.length, totalPassivesB: b.passives.length,
    sharedPassives: diff.passives.shared.length,
    uniquePassivesA: diff.passives.onlyA.length, uniquePassivesB: diff.passives.onlyB.length,
    totalSkillsA: a.skills.length, totalSkillsB: b.skills.length,
    sharedSkills: diff.skills.shared.length,
    uniqueSkillsA: diff.skills.onlyA.length, uniqueSkillsB: diff.skills.onlyB.length,
  };
  return diff;
}

function printDiff(diff) {
  const s = diff.summary;
  const L = (label, val) => `  ${label.padEnd(20)} ${String(val).padStart(6)}`;
  console.log(`\n${'='.repeat(60)}\n  BUILD COMPARISON\n${'='.repeat(60)}\n`);
  console.log(`  Build A: ${diff.buildA.name} (${diff.buildA.ascendancy})`);
  console.log(`  Build B: ${diff.buildB.name} (${diff.buildB.ascendancy})\n`);
  console.log(`  ${'─'.repeat(56)}`);
  console.log(`  ${'SUMMARY'.padStart(30)}\n${'─'.repeat(56)}`);
  console.log(`  ${''.padEnd(20)} ${'A'.padStart(8)} ${'B'.padStart(8)} ${'Shared'.padStart(8)} ${'UniqA'.padStart(8)} ${'UniqB'.padStart(8)}`);
  console.log(`  ${'Passives'.padEnd(20)} ${String(s.totalPassivesA).padStart(8)} ${String(s.totalPassivesB).padStart(8)} ${String(s.sharedPassives).padStart(8)} ${String(s.uniquePassivesA).padStart(8)} ${String(s.uniquePassivesB).padStart(8)}`);
  console.log(`  ${'Skills'.padEnd(20)} ${String(s.totalSkillsA).padStart(8)} ${String(s.totalSkillsB).padStart(8)} ${String(s.sharedSkills).padStart(8)} ${String(s.uniqueSkillsA).padStart(8)} ${String(s.uniqueSkillsB).padStart(8)}\n`);

  if (s.sharedPassives > 0) {
    console.log(`  SHARED PASSIVES (${s.sharedPassives})`);
    for (const p of diff.passives.shared) console.log(`    ${p.id}${p.weaponSet ? ` (WS${p.weaponSet})` : ''}`);
    console.log();
  }

  const printPassives = (items, label) => {
    if (!items.length) return;
    const byCat = {};
    for (const p of items) (byCat[categorizePassive(p.id)] ??= []).push(p);
    console.log(`  ${label}`);
    for (const [cat, list] of Object.entries(byCat).sort()) {
      console.log(`    [${cat}]`);
      for (const p of list) console.log(`      ${p.id}${p.weaponSet ? ` (WS${p.weaponSet})` : ''}`);
    }
    console.log();
  };

  printPassives(diff.passives.onlyA, `UNIQUE TO BUILD A (${diff.buildA.name})`);
  printPassives(diff.passives.onlyB, `UNIQUE TO BUILD B (${diff.buildB.name})`);

  if (s.sharedSkills > 0) {
    console.log(`  SHARED SKILLS (${s.sharedSkills})`);
    for (const sk of diff.skills.shared) {
      console.log(`    ${sk.name}`);
      if (sk.supports.shared.length) console.log(`      Supports: ${sk.supports.shared.join(', ')}`);
      if (sk.supports.onlyA.length) console.log(`      Only in A: ${sk.supports.onlyA.join(', ')}`);
      if (sk.supports.onlyB.length) console.log(`      Only in B: ${sk.supports.onlyB.join(', ')}`);
    }
    console.log();
  }

  const printSkills = (items, label) => {
    if (!items.length) return;
    console.log(`  ${label}`);
    for (const sk of items) {
      console.log(`    ${sk.name}`);
      if (sk.supports?.length) console.log(`      Supports: ${sk.supports.map(ss => ss.name).join(', ')}`);
    }
    console.log();
  };

  printSkills(diff.skills.onlyA, `SKILLS ONLY IN BUILD A (${diff.buildA.name})`);
  printSkills(diff.skills.onlyB, `SKILLS ONLY IN BUILD B (${diff.buildB.name})`);
}

function main() {
  const args = process.argv.slice(2);
  if (args.length !== 2) {
    console.log('Usage: node compare_builds.mjs <build1.build> <build2.build>');
    process.exit(1);
  }
  const a = parseBuild(args[0]);
  const b = parseBuild(args[1]);
  printDiff(compareBuilds(a, b));
}

main();
